### PR TITLE
[enh] Set main domain as hostname

### PR DESCRIPTION
**Pull request by alexAubin**
Attempt to address https://dev.yunohost.org/issues/341, based on https://github.com/labriqueinternet/doctorcube_ynh/pull/9

Still need to think about this and do some tests ... Not entirely sure what to do with /etc/hosts and implications.

### CODE REVIEW

File: src/yunohost/tools.py [hunk 1]
```diff
@@ -162,6 +162,45 @@ def tools_maindomain(auth, new_domain=None):
         logger.warning("%s" % e, exc_info=1)
         raise MoulinetteError(errno.EPERM, m18n.n('maindomain_change_failed'))
 
+    # Set hostname
+    
+    commands = [     
+        "sudo hostnamectl --static set-hostname %s" % new_domain,
+        "sudo hostnamectl --transient set-hostname %s" % new_domain,
+        "sudo hostnamectl --pretty set-hostname (YunoHost/%s)" % new_domain
```
Comment (jvaubourg) [inline src/yunohost/tools.py]: I think you should add quotes (for bash) around the parenthesis of the pretty hostname.

Comment (alexAubin) [inline src/yunohost/tools.py]: I tried before committing, and somehow the quotes ended up in the name itself (in hostnamectl status). This probably has to do with the use of subprocess. Subprocess takes as input a list and each element kind of act like an argument. Not like bash where you put quotes specifically to say "this is to be understood as a single argument" (AFAIK).

Comment (Psycojoker) [inline src/yunohost/tools.py]: You probably want to use the list based approach of subprocess then https://docs.python.org/2/library/subprocess.html#using-the-subprocess-module

**Comment by Psycojoker** [comment]
```
19:43 @<Bram> Aleks: je sais pas s'il faut faire ça dans /etc/hosts
19:43 @<Bram> par contre je vois que ton code est pas assez solide :/
19:43 @<Bram> faudrait utiliser un re
19:44 @<Bram> re.match("127\.0\.0\.1.+ %s"
19:44 @<Bram> un truc du genre
19:44 @<Bram> tu peux avoir plusieurs domains sur une ligne
19:44 @<Bram> et plusieurs espaces/tabs
19:44 @<Bram> c'est assez le merdique
19:45  <Aleks> Oep, c'est un peu la merde de faire un vrai test :/
19:45  <Aleks> Pourrait aussi tenter de resolv le domaine et voir.si ca renvooe 127.0.0.1
```

**Comment by Psycojoker** [comment]
https://github.com/jonhadfield/python-hosts

**Comment by opi** [comment]
is "put the (main) domain(s) in /etc/hosts file" another issue that we need to address in another PR ?

**Comment by Psycojoker** [comment]
Having 2 differents PR would simplify things, indeed.

**Comment by alexAubin** [comment]
Okay, let's stick to only doing this for the main domain for now then.

I'll have a look at providing a better way to parse /etc/hosts as discussed with @Psycojoker.

**Comment by Psycojoker** [comment]
> I'll have a look at providing a better way to parse /etc/hosts as discussed with @Psycojoker.

Regarding this, I think we can actually do something simplier : in neutrinet.sh we only check if the domain is present, then add it and I think that it's totally enough https://github.com/labriqueinternet/configuration_scripts/blob/master/neutrinet.sh#L120

If the domain is present we don't want to add another line that would overwrite the previous one and if it's present it's that the user already did something. We just need a better check than in neutrinet.sh to avoid having false positive result (for example: look for domain.com and being happy because sub.domain.com is already present)

**Comment by alexAubin** [comment]
We could grep on " domain.tld" (with a space before, and a space or end of line after, maybe)

**Comment by Psycojoker** [comment]
Something like `^ *domain.tld` would be better I think (regex)

**Comment by alexAubin** [comment]
We discussed this on the chat with @Psycojoker today, he suggested [this snippet](https://gist.github.com/Psycojoker/a71144e6ea991a88bd66b327a290da0f) to parse /etc/hosts

**Comment by alexAubin** [comment]
Let's postpone this for 2.6, as this is more an improvement than a bug.

**Comment by Psycojoker** [comment]
The `/etc/hosts` modifications should be removed from this PR.

File: src/yunohost/tools.py (hunk 1 above)
Comment (alexAubin) [inline src/yunohost/tools.py]: Reviving this discussion : not sure what to do here. The current code works, and I don't see any benefits of using directly a list (I already do a .split() when using subprocess right after - the only limitation is that we cannot have spaces in the pretty hostname atm, but we don't really care imho)

Comment (Psycojoker) [inline src/yunohost/tools.py]: Looks cleaner and that would avoid stupid bug like "yes you can have a domain name in unicode".

File: src/yunohost/tools.py [hunk 2]
```diff
@@ -162,6 +162,28 @@ def tools_maindomain(auth, new_domain=None):
         logger.warning("%s" % e, exc_info=1)
         raise MoulinetteError(errno.EPERM, m18n.n('maindomain_change_failed'))
 
+    # Set hostname
+    pretty_hostname = "(YunoHost/%s)" % new_domain
```
Comment (opi) [inline src/yunohost/tools.py]: why does "Pretty hostname" prefixed by "Yunohost" ?

Comment (opi) [inline src/yunohost/tools.py]: otherwise everything works fine

Comment (alexAubin) [inline src/yunohost/tools.py]: No particular reason, it's just to have a "pretty" hostname, like not just the FQDN alone but something describing the hostname and machine, if I understand correctly

**Comment by opi** [comment]
please squash before merging this PR to avoir log clutter

**Comment by opi** [comment]
please squash before merging this PR to avoir log clutter

File: src/yunohost/tools.py [hunk 3]
```diff
@@ -162,6 +162,28 @@ def tools_maindomain(auth, new_domain=None):
...
+    commands = [
+        "sudo hostnamectl --static    set-hostname".split() + [new_domain],
+        "sudo hostnamectl --transient set-hostname".split() + [new_domain],
+        "sudo hostnamectl --pretty    set-hostname".split() + [pretty_hostname]
```
Comment (zamentur) [inline src/yunohost/tools.py]: I 'm not sure this kind of commands are persistent. And you ?

Comment (alexAubin) [inline src/yunohost/tools.py]: I think hostnamectl is precisely designed to make change persitent (unlike hostname and other ways). Also this is already what Doctor Cube does : https://github.com/labriqueinternet/doctorcube_ynh/blob/6d8580c0216c925cdd2ea435a23a9f0408f6a471/scripts/install#L18-L20

**Pull request merged**